packages/block/src/header.ts

```diff
@@ -351,210 +337,38 @@ export class BlockHeader {
-   *
-   * @param parentBlockHeader - the header from the parent `Block` of this header
-   */
-  canonicalDifficulty(parentBlockHeader: BlockHeader): bigint {
-    if (this._common.consensusType() !== ConsensusType.ProofOfWork) {
-      const msg = this._errorMsg('difficulty calculation is only supported on PoW chains')
-      throw new Error(msg)
-    }
-    if (this._common.consensusAlgorithm() !== ConsensusAlgorithm.Ethash) {
-      const msg = this._errorMsg(
-        'difficulty calculation currently only supports the ethash algorithm'
-      )
+    // check if the block used too much gas
+    if (this.gasUsed > this.gasLimit) {
+      const msg = this._errorMsg('Invalid block: too much gas used')
       throw new Error(msg)
     }
-    const hardfork = this._common.hardfork()
-    const blockTs = this.timestamp
-    const { timestamp: parentTs, difficulty: parentDif } = parentBlockHeader
-    const minimumDifficulty = this._common.paramByHardfork('pow', 'minimumDifficulty', hardfork)
-    const offset =
-      parentDif / this._common.paramByHardfork('pow', 'difficultyBoundDivisor', hardfork)
-    let num = this.number
-
-    // We use a ! here as TS cannot follow this hardfork-dependent logic, but it always gets assigned
-    let dif!: bigint
-
-    if (this._common.hardforkGteHardfork(hardfork, Hardfork.Byzantium)) {
-      // max((2 if len(parent.uncles) else 1) - ((timestamp - parent.timestamp) // 9), -99) (EIP100)
-      const uncleAddend = parentBlockHeader.uncleHash.equals(KECCAK256_RLP_ARRAY) ? 1 : 2
-      let a = BigInt(uncleAddend) - (blockTs - parentTs) / BigInt(9)
-      const cutoff = BigInt(-99)
-      // MAX(cutoff, a)
-      if (cutoff > a) {
-        a = cutoff
-      }
-      dif = parentDif + offset * a
-    }
 
-    if (this._common.hardforkGteHardfork(hardfork, Hardfork.Byzantium)) {
-      // Get delay as parameter from common
-      num = num - this._common.param('pow', 'difficultyBombDelay')
-      if (num < BigInt(0)) {
-        num = BigInt(0)
-      }
-    } else if (this._common.hardforkGteHardfork(hardfork, Hardfork.Homestead)) {
-      // 1 - (block_timestamp - parent_timestamp) // 10
-      let a = BigInt(1) - (blockTs - parentTs) / BigInt(10)
-      const cutoff = BigInt(-99)
-      // MAX(cutoff, a)
-      if (cutoff > a) {
-        a = cutoff
+    // Validation for EIP-1559 blocks
+    if (this._common.isActivatedEIP(1559)) {
+      if (typeof this.baseFeePerGas !== 'bigint') {
+        const msg = this._errorMsg('EIP1559 block has no base fee field')
+        throw new Error(msg)
       }
-      dif = parentDif + offset * a
-    } else {
-      // pre-homestead
-      if (parentTs + this._common.paramByHardfork('pow', 'durationLimit', hardfork) > blockTs) {
-        dif = offset + parentDif
-      } else {
-        dif = parentDif - offset
+      const londonHfBlock = this._common.hardforkBlock(Hardfork.London)
+      const isInitialEIP1559Block = londonHfBlock && this.number === londonHfBlock
```

Comment:
Since we know that this.number is defined, perhaps we could simplify this a bit:
```suggestion
      const isInitialEIP1559Block = this.number === this._common.hardforkBlock(Hardfork.London)
```